### ISSUE TITLE
Add a way for this container to be used for development 

### ISSUE THREAD

**Issue by atrauzzi** [aspnet/aspnet-docker#101]
It would be neat if perhaps in some way, this container could be used by all the IDEs for compilation and syntax checking/code assist instead of requiring a local installation.

**Comment by muratg** [comment]
This sounds like an interesting idea, though we probably won't get to this before V1 actually ships.

**Comment by atrauzzi** [comment]
Definitely.  I should add: Tooling support is fairly important here.

**Comment by atrauzzi** [comment]
Created a user voice submission for VS code: https://visualstudio.uservoice.com/forums/293070-visual-studio-code/suggestions/10682943-allow-containerized-dnx-runtimes-to-be-used-for-de

**Comment by muratg** [comment]
Thanks @atrauzzi! The way I envision this is as a community-driven project, involving OmniSharp and a docker image based on the official one.

**Comment by atrauzzi** [comment]
I've made #116 and #113 which likely fall under everything here.

What might be nice is if a discussion could be started somewhere about how this stuff could all work.  Indeed I think @omnisharp should be involved as there are some more things (debugging dnx running in a container) that need to be covered.  Also, whether a dev mode should be done by the container, or if dev-mode functionality is better handled by the omnisharp tooling recognizing changes and firing commands into the container...

**Comment by AlexChesser** [comment]
Hey Alex, I've been doing some stuff that might help here. Although it is not actually working, I may be able to take another pass at it by the weekend. I'd love to hear your thoughts.

I've been starting DEV locally with a local docker compose file as you see here:

https://github.com/AlexChesser/aspnet-docker-sample/blob/master/docker-compose.yml

By creating the mount-to-src point (in theory) the code is compiled and running in the container, but "against" the files that live on the host OS. I've found that this *actually* works in a nodejs / express app.

Unfortunately, it's mildly broken with DNX because (I'm speculating) the nuget restore in the docker file can't see the code in the src-folder during the `docker build` phase since the volume doesn't exist until later in the docker life cycle.

I'm still very new to building docker files so don't know everything that can and can't be done... But if the docker-file was something like

```
FROM microsoft/aspnet
COPY . /app
WORKDIR /app

RUN ["dnu","restore"]

EXPOSE 80
EXPOSE 443
EXPOSE 5000

#run the Linux equivalent of Robocopy with the "mirror" and "listen" flags to
# automatically copy from the /SRC folder to the /app folder (SRC was mounted as part of docker compose)


# need a bit more magic to handle the recompile and reload commands, as it stands the 
# 'Robocopy' above should handle CSS, HTML, and SASS (if grunt watch is running)
ENTRYPOINT ["dnx", "-p", "project.json", "web"]
```

What do you think? Sound promising?

**Comment by atrauzzi** [comment]
@AlexChesser - Glad to hear I'm not the only Alex interested in seeing this made!  ;)

While I'm not an expert on ASP.NET, this is the list so far of considerations that I've thought up based on my experience building php-fpm container setups:

 - COPY/ADD the project directory to `/app` during build.
 - During development, bind mount the project directory to `/app`, which overrides the `COPY`/`ADD`ed project for that container instance.  This would obviate the need to use an additional tool like robocopy
   - Remember however that the bind mount will have the UID, GID and permissions of the filesystem it originates from.  There may be dragons there.
 - The user the container uses when running commands is most likely root, and the user the container uses when running the actual process could be different to that as well.  More dragons.
 - Be mindful of the fact that `dnx` (and soon dotnet) installs dependencies to a cache, not as part of the project file structure.  The implication here is that all your dependencies will likely be in `/root/.wherever` *inside the container*.
   - This might mean that you want to also blank out that path during development as well so that dependencies from development aren't permanently written to the running container
 - Some port will likely need to be exposed to allow a debugger to connect, also, runtime parameters when the container is started will likely be needed to toggle the debugger on and off

The main thing, which we've both eluded to in our previous posts are:

 - Restarting the server when code changes
   - Currently I lean towards having the tooling recognize this on the native system rather than filling the container up with functionality that doesn't pertain to what it's doing.  Some kind of dev-time tooling would know how to watch for local file changes and trigger commands inside the running server container to do "stuff" (restart, dep updates)
   - I'm a bit confused and maybe an ASP.NET expert can chime in, but I was always under the impression that there was support for dynamic recompilation.  I'm not sure where that falls in this brave new world for .NET though
 - It would be nice if any filesystem watching functionality could also support watching files to run other tools like NPM and JSPM inside the container

On top of a `Dockerfile`, we'll probably also need to ship a `docker-compose.yml`.  Then when people want to do local development, they just have to go `docker-compose up` and `docker-compose run` to talk into the container.  Again, tooling support is really what will make this easy, so that say omnisharp could know that the project is running in a container and connect to it for debugging.

That's I think the extent of my daydreaming on the subject.

tldr; With all the above considered, you'd be able to customize the `Dockerfile` and `docker-compose.yml` in an asp.net project to represent your desired architecture and also include your front end build stack.  Your editor (VS Code, omnisharp) would then just be calling commands in the container based on watched files and connecting a debugger when needed.
